### PR TITLE
feat(txpool): add best_with_base_fee

### PR DESCRIPTION
**Pull request by mattsse**
extends the `best` function with a basfee parameter that returns all pending transactions that satisfy the basfee.

this essentially moves unlocked txs from basefee pool to pending iterator if basefee decreased
and filters invalid transactions in the best iterator

This is useful for block building which need to enforce a certain basefee, and it could be that the pool isn't updated yet when a block building job is triggered

### CODE REVIEW

File: crates/transaction-pool/src/pool/best.rs [hunk 1]
```diff
@@ -10,6 +10,40 @@ use std::{
...
+///
+/// This iterator guarantees that all transaction it returns satisfy the base fee.
+pub(crate) struct BestTransactionsWithBasefee<T: TransactionOrdering> {
+    pub(crate) best: BestTransactions<T>,
+    pub(crate) base_fee: u128,
+}
+
+impl<T: TransactionOrdering> crate::traits::BestTransactions for BestTransactionsWithBasefee<T> {
+    fn mark_invalid(&mut self, tx: &Self::Item) {
+        BestTransactions::mark_invalid(&mut self.best, tx)
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/best.rs:26]: nit: `self.best.mark_invalid(tx)` should work too?

File: crates/transaction-pool/src/pool/best.rs [hunk 2]
```diff
@@ -10,6 +10,40 @@ use std::{
...
+impl<T: TransactionOrdering> Iterator for BestTransactionsWithBasefee<T> {
+    type Item = Arc<ValidPoolTransaction<T::Transaction>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // find the next transaction that satisfies the base fee
+        loop {
+            let best = self.best.next()?;
+            if best.transaction.max_fee_per_gas() < self.base_fee {
+                // tx violates base fee, mark it as invalid and continue
+                crate::traits::BestTransactions::mark_invalid(self, &best);
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/best.rs:39]:
```suggestion
                self.mark_invalid(&best);
```

File: crates/transaction-pool/src/pool/best.rs [hunk 3]
```diff
@@ -10,6 +10,40 @@ use std::{
...
+///
+/// This is a wrapper around [`BestTransactions`] that also enforces a specific basefee.
+///
+/// This iterator guarantees that all transaction it returns satisfy the base fee.
+pub(crate) struct BestTransactionsWithBasefee<T: TransactionOrdering> {
+    pub(crate) best: BestTransactions<T>,
+    pub(crate) base_fee: u128,
+}
+
+impl<T: TransactionOrdering> crate::traits::BestTransactions for BestTransactionsWithBasefee<T> {
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/best.rs:24]: import the trait at top of file

File: crates/transaction-pool/src/traits.rs [hunk 4]
```diff
@@ -175,6 +175,15 @@ pub trait TransactionPool: Send + Sync + Clone {
...
     ) -> Box<dyn BestTransactions<Item = Arc<ValidPoolTransaction<Self::Transaction>>>>;
 
+    /// Returns an iterator that yields transactions that are ready for block production with the
+    /// given base fee.
+    ///
+    /// Consumer: Block production
+    fn best_transactions_with_base_fee(
+        &self,
+        base_fee: u128,
+    ) -> Box<dyn BestTransactions<Item = Arc<ValidPoolTransaction<Self::Transaction>>>>;
```
Comment (gakonst) [inline crates/transaction-pool/src/traits.rs:185]: Starting to worry a little (not due to this PR) about the number of transaction types we have across:
* DB
* TXPool (ValidPoolTransaction, PendingTransaction{Ref} etc.)
* RPC
* ..?

We should document these and during the offsite understand what the overlap looks like/and how you go from one to the other.

File: crates/transaction-pool/src/pool/txpool.rs [hunk 5]
```diff
@@ -186,6 +186,31 @@ impl<T: TransactionOrdering> TxPool<T> {
...
+            }
+            Ordering::Greater => {
+                // base fee increased, we only need to enforces this on the pending pool
+                Box::new(self.pending_pool.best_with_basefee(basefee))
+            }
+            Ordering::Less => {
+                // base fee decreased, we need to move transactions from the basefee pool to the
+                // pending pool
+                let unlocked = self.basefee_pool.satisfy_base_fee_transactions(basefee);
+                Box::new(self.pending_pool.best_with_unlocked(unlocked))
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/txpool.rs:209]: 👍

File: crates/transaction-pool/src/pool/pending.rs [hunk 6]
```diff
@@ -84,6 +85,43 @@ impl<T: TransactionOrdering> PendingPool<T> {
...
+            let transaction = PendingTransactionRef { submission_id, transaction: tx, priority };
+            if best.ancestor(&tx_id).is_none() {
+                best.independent.insert(transaction.clone());
+            }
+            let transaction = Arc::new(PendingTransaction { transaction });
+            best.all.insert(tx_id, transaction);
+        }
+
+        best
+    }
```
Comment (gakonst) [inline crates/transaction-pool/src/pool/pending.rs:123]: needs follow-up good-first-issues for unit tests

**Review by gakonst — APPROVED** [review]

File: crates/transaction-pool/src/pool/best.rs (hunk 1 above)
Comment (mattsse) [reply crates/transaction-pool/src/pool/best.rs:26]: the impl type has the as name I can perhaps make this nicer though

File: crates/transaction-pool/src/traits.rs (hunk 4 above)
Comment (mattsse) [reply crates/transaction-pool/src/traits.rs:185]: I can write a primer and put it in the docs or in rpc

Comment (gakonst) [reply crates/transaction-pool/src/traits.rs:185]: sg! thx!

**Pull request merged**